File: ascii_fields/animations/chladni.py
import math

from ..core import Animation, clamp, render_field


# Chladni mode pairs for a square plate (free edges, classical sin sin difference).
MODES = ((2, 3), (3, 4), (4, 5), (5, 6), (4, 7), (3, 5), (6, 7), (2, 5), (5, 3), (7, 4))
MODE_SECONDS = 4.5
TRANSITION = 1.4


class ChladniAnimation(Animation):
  """A square plate vibrated at one of its eigenfrequencies. The standing wave
  u(x, y) = sin(m pi x) sin(n pi y) - sin(n pi x) sin(m pi y) has nodal lines
  where the plate isn't moving -- sand piles up there. The driving frequency
  slowly steps through the (m, n) modes, drawing different patterns."""

  thresholds = [
    (0.10, " "), (0.22, "."), (0.34, ":"), (0.46, "-"), (0.58, "="),
    (0.70, "+"), (0.82, "*"), (0.92, "#"), (1.01, "@"),
  ]

  def render(self, width, height, elapsed, phase, options):
    idx = int(elapsed / MODE_SECONDS) % len(MODES)
    nxt = (idx + 1) % len(MODES)
    frac = (elapsed % MODE_SECONDS) / MODE_SECONDS
    blend = clamp((frac - (1.0 - TRANSITION / MODE_SECONDS)) / (TRANSITION / MODE_SECONDS))
    m0, n0 = MODES[idx]
    m1, n1 = MODES[nxt]
    omega = 1.6
    # The sand sits at nodal lines regardless of how hard the plate is shaken,
    # so the *pattern* is amplitude-independent; the drive only pulses the
    # overall brightness slightly so the scene feels alive.
    pulse = 0.78 + 0.22 * abs(math.cos(omega * elapsed))
    contrast = options.contrast
    grid = []
    for row in range(height):
      y = row / max(1, height - 1)
      sin0y_m0 = math.sin(m0 * math.pi * y)
      sin0y_n0 = math.sin(n0 * math.pi * y)
      sin1y_m1 = math.sin(m1 * math.pi * y)
      sin1y_n1 = math.sin(n1 * math.pi * y)
      line = []
      for col in range(width):
        x = col / max(1, width - 1)
        u0 = math.sin(m0 * math.pi * x) * sin0y_n0 - math.sin(n0 * math.pi * x) * sin0y_m0
        u1 = math.sin(m1 * math.pi * x) * sin1y_n1 - math.sin(n1 * math.pi * x) * sin1y_m1
        u = u0 * (1.0 - blend) + u1 * blend
        # sand collects at the nodal lines |u| = 0; sharper exp -> tighter lines
        nodal = math.exp(-abs(u) * 6.0)
        line.append(clamp(nodal * pulse * contrast))
      grid.append(line)
    return render_field(width, height, grid, options, self)

File: src/animations/chladni.rs
use crate::animation::{Animation, FrameContext};
use crate::core::{clamp, render_field, FieldStyle};
use std::f64::consts::PI;

const STYLE: FieldStyle = FieldStyle { gray_lo: 234, gray_hi: 255, default_theme: "amber" };
const TH: &[(f64, char)] = &[
  (0.10, ' '),
  (0.22, '.'),
  (0.34, ':'),
  (0.46, '-'),
  (0.58, '='),
  (0.70, '+'),
  (0.82, '*'),
  (0.92, '#'),
  (1.01, '@'),
];
const MODES: &[(f64, f64)] = &[
  (2.0, 3.0),
  (3.0, 4.0),
  (4.0, 5.0),
  (5.0, 6.0),
  (4.0, 7.0),
  (3.0, 5.0),
  (6.0, 7.0),
  (2.0, 5.0),
  (5.0, 3.0),
  (7.0, 4.0),
];
const MODE_SECONDS: f64 = 4.5;
const TRANSITION: f64 = 1.4;

pub struct Chladni;

impl Animation for Chladni {
  fn render(&mut self, ctx: &FrameContext, out: &mut String) {
    let n_modes = MODES.len();
    let idx = ((ctx.elapsed / MODE_SECONDS) as usize) % n_modes;
    let nxt = (idx + 1) % n_modes;
    let frac = (ctx.elapsed % MODE_SECONDS) / MODE_SECONDS;
    let blend = clamp((frac - (1.0 - TRANSITION / MODE_SECONDS)) / (TRANSITION / MODE_SECONDS));
    let (m0, n0) = MODES[idx];
    let (m1, n1) = MODES[nxt];
    let omega = 1.6;
    let pulse = 0.78 + 0.22 * (omega * ctx.elapsed).cos().abs();
    let contrast = ctx.options.contrast;
    let w = ctx.width;
    let h = ctx.height;
    let mut grid = vec![0.0_f64; w * h];
    let dw = (w.saturating_sub(1)).max(1) as f64;
    let dh = (h.saturating_sub(1)).max(1) as f64;
    for row in 0..h {
      let y = row as f64 / dh;
      let sin0y_m0 = (m0 * PI * y).sin();
      let sin0y_n0 = (n0 * PI * y).sin();
      let sin1y_m1 = (m1 * PI * y).sin();
      let sin1y_n1 = (n1 * PI * y).sin();
      let base = row * w;
      for col in 0..w {
        let x = col as f64 / dw;
        let u0 = (m0 * PI * x).sin() * sin0y_n0 - (n0 * PI * x).sin() * sin0y_m0;
        let u1 = (m1 * PI * x).sin() * sin1y_n1 - (n1 * PI * x).sin() * sin1y_m1;
        let u = u0 * (1.0 - blend) + u1 * blend;
        let nodal = (-u.abs() * 6.0).exp();
        grid[base + col] = clamp(nodal * pulse * contrast);
      }
    }
    render_field(ctx, &grid, TH, &STYLE, out);
  }
}
